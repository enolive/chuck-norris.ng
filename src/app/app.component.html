<h1>{{title}}</h1>
<img src="../assets/chuck-norris-logo.jpg" alt="Chuck Norris Approves!">
<p role="status">{{jokeText}}</p>
<button (click)="refreshJoke()">Refresh</button>
