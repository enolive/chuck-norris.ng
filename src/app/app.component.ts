import {Component, OnInit} from '@angular/core';
import {HttpClient} from '@angular/common/http';
import {map} from 'rxjs/operators';

interface JokeResponse {
  value: {
    joke: string
  };
}

@Component({
  selector: 'app-root',
  templateUrl: './app.component.html',
  styleUrls: ['./app.component.css']
})
export class AppComponent implements OnInit {
  title = 'Chuck Norris';
  jokeText = 'Chuck Norris Joke';

  constructor(private httpClient: HttpClient) {
  }

  refreshJoke() {
    const replaceHtmlEntities = (text: string) => text.replace(/&quot;/g, '"');
    this.httpClient
        .get<JokeResponse>('http://api.icndb.com/jokes/random')
        .pipe(
          map(response => response.value.joke),
          map(replaceHtmlEntities))
        .subscribe(value => {
          this.jokeText = value;
        });
  }

  ngOnInit(): void {
    this.refreshJoke();
  }
}
