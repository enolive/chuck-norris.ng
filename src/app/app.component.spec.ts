import {AppComponent} from './app.component';
import {fireEvent, render} from '@testing-library/angular';
import {HttpClient, HttpClientModule} from '@angular/common/http';
import {createMock} from '@testing-library/angular/jest-utils';
import {of} from 'rxjs';

describe('AppComponent', () => {
  test('is rendered', async () => {
    const component = await render(AppComponent, {
      componentProperties: {
        title: 'Chuck Norris',
        jokeText: 'Random Joke'
      },
      imports: [
        HttpClientModule,
      ]
    });

    expect(component.container).toMatchSnapshot();
  });

  test('refreshes joke on click', async () => {
    const expectedJoke = 'Chuck Norris uses canvas in IE.';
    const http = createMock(HttpClient);
    http.get = jest.fn().mockReturnValue(of({
      type: 'success',
      value: {id: 535, joke: expectedJoke, categories: ['nerdy']}
    }));
    const component = await render(AppComponent, {
      providers: [
        {provide: HttpClient, useValue: http},
      ]
    });

    const button = component.getByRole('button');
    fireEvent.click(button);

    expect(http.get).toHaveBeenCalledTimes(2);
    expect(http.get).toHaveBeenCalledWith('http://api.icndb.com/jokes/random');
    expect(component.fixture.componentInstance.jokeText).toBe(expectedJoke);
    expect(component.getByRole('status')).toHaveTextContent(expectedJoke);
  });

  test('refreshes joke on init', async () => {
    const http = createMock(HttpClient);
    http.get = jest.fn().mockReturnValue(of({value: {joke: 'joke'}}));
    await render(AppComponent, {
      providers: [
        {provide: HttpClient, useValue: http},
      ]
    });
    expect(http.get).toHaveBeenCalledTimes(1);
  });

  test('renders quotes from joke', async () => {
    const http = createMock(HttpClient);
    http.get = jest.fn().mockReturnValue(of({value: {joke: '&quot;&quot;&quot;'}}));
    const component = await render(AppComponent, {
      providers: [
        {provide: HttpClient, useValue: http},
      ]
    });
    expect(component.fixture.componentInstance.jokeText).toBe('"""');
  });
});
