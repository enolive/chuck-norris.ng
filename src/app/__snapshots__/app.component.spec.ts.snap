// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppComponent is rendered 1`] = `
<div
  id="root0"
>
  <h1>
    Chuck Norris
  </h1>
  <img
    alt="Chuck Norris Approves!"
    src="../assets/chuck-norris-logo.jpg"
  />
  <p
    role="status"
  >
    Random Joke
  </p>
  <button>
    Refresh
  </button>
</div>
`;
